haskell-src/db-schema/migrations/2.1.1_transactions_pactid_index.sql

```diff
@@ -0,0 +1,3 @@
+CREATE INDEX transactions_pactid_index 
+  ON transactions (pactid, goodresult DESC, height DESC) 
```

Comment:
Including `goodresult` in the index like this isn't very good for performance since it's an arbitrarily long text, but even more important than performance, [it will cause errors](https://dba.stackexchange.com/questions/25138/index-max-row-size-error) if we try to insert a row with `pactid IS NOT NULL` and `goodresult` is larger than a few kilobytes. I've run the following query and it turned out that the largest `goodresult` in the database is ~300kb!
``` SQL
SELECT length(goodresult::text), goodresult FROM transactions WHERE goodresult IS NOT NULL ORDER BY length(goodresult::text) DESC LIMIT 1;
```
I think we should instead index on `goodresult IS NOT NULL`

Reply:
While also including the later suggestion to not join on the blocks table, I've updated the index above to not include all of `goodresult` (just a predicate on it). Here is the query plan.
```
                                                                               QUERY PLAN                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=876.78..26368.20 rows=30 width=1963) (actual time=0.022..0.023 rows=0 loops=1)
   ->  Index Scan using transactions_pactid_index2 on transactions t0  (cost=0.42..76.72 rows=30 width=1899) (actual time=0.021..0.022 rows=0 loops=1)
         Index Cond: (((pactid)::text = 'zrkQoQkwClQLysRH0Ulqj92-zOYQC9x833DsEQPauFc'::text) AND ((goodresult IS NOT NULL) = true))
   ->  Aggregate  (cost=876.35..876.36 rows=1 width=64) (never executed)
         CTE transactionsteps
           ->  Recursive Union  (cost=0.56..873.82 rows=101 width=282) (never executed)
                 ->  Unique  (cost=0.56..8.57 rows=1 width=282) (never executed)
                       ->  Index Scan using transactions_requestkey_idx on transactions tinner  (cost=0.56..8.57 rows=1 width=282) (never executed)
                             Index Cond: ((requestkey)::text = ($1)::text)
                 ->  Subquery Scan on "*SELECT* 2"  (cost=86.27..86.42 rows=10 width=282) (never executed)
                       ->  Unique  (cost=86.27..86.32 rows=10 width=286) (never executed)
                             ->  Sort  (cost=86.27..86.30 rows=11 width=286) (never executed)
                                   Sort Key: trec.depth
                                   ->  Nested Loop  (cost=0.56..86.08 rows=11 width=286) (never executed)
                                         ->  WorkTable Scan on transactionsteps trec  (cost=0.00..0.20 rows=10 width=36) (never executed)
                                         ->  Index Scan using transactions_requestkey_idx on transactions tinner_1  (cost=0.56..8.57 rows=1 width=278) (never executed)
                                               Index Cond: ((requestkey)::text = (trec.pactid)::text)
         ->  CTE Scan on transactionsteps  (cost=0.00..2.02 rows=101 width=68) (never executed)
 Planning Time: 0.331 ms
 Execution Time: 0.083 ms
```
Note that we use an index `transactions_pactid_index2`, which is an index I temporarily created. It has the following expected definition.
```
 CREATE INDEX transactions_pactid_index2 ON public.transactions USING btree (pactid, ((goodresult IS NOT NULL)), height DESC) WHERE (pactid IS NOT NULL)
```

Reply:
I made a slight error in this previous query. The commit (58118184) includes a change that doesn't exclude rows with a null `goodResult` value (transactions that did not succeed), but instead we sort the results so that the successful transactions show up first. Here is the resulting query plan. Look at the above comment for an explanation of the index `transactions_pactid_index2`
```
                                                                                               QUERY PLAN                                         
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=30819.20..30819.29 rows=35 width=1964) (actual time=0.236..0.238 rows=16 loops=1)
   Sort Key: ((t0.goodresult IS NOT NULL)) DESC, t0.height DESC
   Sort Method: quicksort  Memory: 41kB
   ->  Nested Loop  (cost=876.78..30818.30 rows=35 width=1964) (actual time=0.067..0.218 rows=16 loops=1)
         ->  Index Scan using transactions_pactid_index2 on transactions t0  (cost=0.42..144.91 rows=35 width=1899) (actual time=0.019..0.028 rows=16 loops=1)
               Index Cond: ((pactid)::text = 'zrkQoQkwClQLysRH0Ulqj92-zOYQC9x833DsEQPauFc'::text)
         ->  Aggregate  (cost=876.35..876.36 rows=1 width=64) (actual time=0.011..0.011 rows=1 loops=16)
               CTE transactionsteps
                 ->  Recursive Union  (cost=0.56..873.82 rows=101 width=282) (actual time=0.005..0.008 rows=1 loops=16)
                       ->  Unique  (cost=0.56..8.57 rows=1 width=282) (actual time=0.005..0.006 rows=1 loops=16)
                             ->  Index Scan using transactions_requestkey_idx on transactions tinner  (cost=0.56..8.57 rows=1 width=282) (actual time=0.005..0.005 rows=1 loops=16)
                                   Index Cond: ((requestkey)::text = ($1)::text)
                       ->  Subquery Scan on "*SELECT* 2"  (cost=86.27..86.42 rows=10 width=282) (actual time=0.002..0.002 rows=0 loops=16)
                             ->  Unique  (cost=86.27..86.32 rows=10 width=286) (actual time=0.002..0.002 rows=0 loops=16)
                                   ->  Sort  (cost=86.27..86.30 rows=11 width=286) (actual time=0.002..0.002 rows=0 loops=16)
                                         Sort Key: trec.depth
                                         Sort Method: quicksort  Memory: 25kB
                                         ->  Nested Loop  (cost=0.56..86.08 rows=11 width=286) (actual time=0.001..0.001 rows=0 loops=16)
                                               ->  WorkTable Scan on transactionsteps trec  (cost=0.00..0.20 rows=10 width=36) (actual time=0.000..0.000 rows=1 loops=16)
                                               ->  Index Scan using transactions_requestkey_idx on transactions tinner_1  (cost=0.56..8.57 rows=1 width=278) (actual time=0.000..0.000 rows=0 loops=16)
                                                     Index Cond: ((requestkey)::text = (trec.pactid)::text)
               ->  CTE Scan on transactionsteps  (cost=0.00..2.02 rows=101 width=68) (actual time=0.006..0.008 rows=1 loops=16)
 Planning Time: 0.482 ms
 Execution Time: 0.306 ms
```

Reply:
Thanks for the information and the query plans, would you mind sharing the query itself along with the query plan. No need to format the query.

Reply:
Thanks for the query. I see 2 problems here:
1) The query doesn't have any `LIMIT` clause, seems like we're ignoring the `limit` parameter of the HTTP request and not applying any limit ourselves either, so this endpoint could return thousands of rows for certain `pactid`s.
2) Looking at the query plan, I see that Postgres is deciding to do the sorting after scanning the `transactions_pactid_index2` even though `transactions_pactid_index2` is already sorted in the same way as this query's `ORDER BY`. The trouble here is that if it needs to do the sorting after a full-scan, it will have to visit all the rows, even though it needs `LIMIT` many rows (once we have the `LIMIT` clause).